In [25]:
# Import the required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path

## Import the data 

In [26]:
# Using the read_csv function and Path module, create a DataFrame 
# by importing the sfo_neighborhoods_census_data.csv file from the Resources folder
sfo_data_csv = Path('Resources/sfo_neighborhoods_census_data.csv')
sfo_data_df = pd.read_csv(sfo_data_csv)

#clean dataframes of any nulls and drop duplicates

sfo_data_df.isnull()
sfo_data_df.dropna(inplace=True)
 
sfo_data_df.isnull().sum()




year                   0
neighborhood           0
sale_price_sqr_foot    0
housing_units          0
gross_rent             0
dtype: int64

In [27]:
#Check for duplicated rpws om columns 
sfo_data_df.duplicated()


0      False
1      False
2      False
3      False
4      False
       ...  
392    False
393    False
394    False
395    False
396    False
Length: 392, dtype: bool

In [28]:
sfo_data_df.drop_duplicates(subset=['year', 'neighborhood', 'sale_price_sqr_foot', 'housing_units', 'gross_rent'])

year            neighborhood  sale_price_sqr_foot  housing_units  \
0    2010            Alamo Square           291.182945         372560   
1    2010              Anza Vista           267.932583         372560   
2    2010                 Bayview           170.098665         372560   
3    2010        Buena Vista Park           347.394919         372560   
4    2010        Central Richmond           319.027623         372560   
..    ...                     ...                  ...            ...   
392  2016          Telegraph Hill           903.049771         384242   
393  2016              Twin Peaks           970.085470         384242   
394  2016  Van Ness/ Civic Center           552.602567         384242   
395  2016       Visitacion Valley           328.319007         384242   
396  2016           Westwood Park           631.195426         384242   

     gross_rent  
0          1239  
1          1239  
2          1239  
3          1239  
4          1239  
..          ...  
392        4390  
393        4390  
394        4390  
395        4390  
396        4390  

[392 rows x 5 columns]

In [29]:
#Review the first five rows of the DataFrame
#sfo_data_df.head(5)
display(sfo_data_df)

year            neighborhood  sale_price_sqr_foot  housing_units  \
0    2010            Alamo Square           291.182945         372560   
1    2010              Anza Vista           267.932583         372560   
2    2010                 Bayview           170.098665         372560   
3    2010        Buena Vista Park           347.394919         372560   
4    2010        Central Richmond           319.027623         372560   
..    ...                     ...                  ...            ...   
392  2016          Telegraph Hill           903.049771         384242   
393  2016              Twin Peaks           970.085470         384242   
394  2016  Van Ness/ Civic Center           552.602567         384242   
395  2016       Visitacion Valley           328.319007         384242   
396  2016           Westwood Park           631.195426         384242   

     gross_rent  
0          1239  
1          1239  
2          1239  
3          1239  
4          1239  
..          ...  
392        4390  
393        4390  
394        4390  
395        4390  
396        4390  

[392 rows x 5 columns]

In [30]:
# Using the read_csv function and Path module, create a DataFrame 
# Import neighbourbood_coordinates_csv
nh_coordinates_csv = Path('Resources/neighborhoods_coordinates.csv')
nh_coordinates_df = pd.read_csv(nh_coordinates_csv)
#clean dataframes of any nulls 
nh_coordinates_df.dropna(inplace=True)

In [31]:
nh_coordinates_df.head()

Neighborhood        Lat         Lon
0     Alamo Square  37.791012 -122.402100
1       Anza Vista  37.779598 -122.443451
2          Bayview  37.734670 -122.401060
3  Bayview Heights  37.728740 -122.410980
4   Bernal Heights  37.728630 -122.443050

In [32]:
nh_coordinates_df.tail()

Neighborhood       Lat         Lon
68         West Portal  37.74026 -122.463880
69    Western Addition  37.79298 -122.435790
70  Westwood Highlands  37.73470 -122.456854
71       Westwood Park  37.73415 -122.457000
72         Yerba Buena  37.79298 -122.396360

In [33]:
housing_peryear_csv = Path('Resources/housing_per_year.csv')
housing_peryear_df = pd.read_csv(housing_peryear_csv)
housing_peryear_df.dropna(inplace=True)
housing_peryear_df.drop_duplicates()
housing_peryear_df


2010  372560
0  2011  374507
1  2012  376454
2  2013  378401
3  2014  380348
4  2015  382295
5  2016  384242

---

### Step 1: Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

In [34]:
# Create a numerical aggregation that groups the data by the year and then averages the results.
housing_units_by_year = round(sfo_data_df.groupby('year').mean())

# Review the DataFrame
housing_units_by_year

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010                369.0       372560.0      1239.0
2011                342.0       374507.0      1530.0
2012                399.0       376454.0      2324.0
2013                484.0       378401.0      2971.0
2014                556.0       380348.0      3528.0
2015                633.0       382295.0      3739.0
2016                698.0       384242.0      4390.0

### Step 2: Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

### Step 3: Style and format the line plot to ensure a professionally styled visualization.

In [35]:
# Create a visual aggregation explore the housing units by year
#housing_units_by_year.hvplot.bar(x='year', y='housing_units',ylim=[350000, 400000])

housing_units_by_year.hvplot.bar(
    x='year',
    y='housing_units',
    rot=90,
    title='Housing Units In San Fransisco From 2010-2016',
    xlabel = 'year',
    ylable='housing units',
    ylim=[350000, 400000]
).opts(yformatter='%0f')

:Bars   [year]   (housing_units)

### Step 5: Answer the following question:

**Question:** What is the overall trend in housing_units over the period being analyzed?

**Answer:** # The overal trend of the housing units have stayed almost consistent over the years. This only goes to sheow that development of units have been steady. 

---

### Step 1: Group the data by year, and then average the results.

In [36]:
# Create a numerical aggregation by grouping the data by year and averaging the results
prices_square_foot_by_year= round(sfo_data_df.groupby('year').mean()) 

# Review the resulting DataFrame
#prices_square_foot_by_data.min()
prices_square_foot_by_year

sale_price_sqr_foot  housing_units  gross_rent
year                                                
2010                369.0       372560.0      1239.0
2011                342.0       374507.0      1530.0
2012                399.0       376454.0      2324.0
2013                484.0       378401.0      2971.0
2014                556.0       380348.0      3528.0
2015                633.0       382295.0      3739.0
2016                698.0       384242.0      4390.0

**Question:** What is the lowest gross rent reported for the years included in the DataFrame?

**Answer:** # The lowest gross rent reported for the years, would be 2010 for the amount of $1239.00

### Step 2: Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

In [37]:
# Filter out the housing_units column, creating a new DataFrame 
# Keep only sale_price_sqr_foot and gross_rent averages per year
prices_square_foot_by_year = prices_square_foot_by_year.drop(columns=['housing_units'])

# Review the DataFrame
prices_square_foot_by_year

sale_price_sqr_foot  gross_rent
year                                 
2010                369.0      1239.0
2011                342.0      1530.0
2012                399.0      2324.0
2013                484.0      2971.0
2014                556.0      3528.0
2015                633.0      3739.0
2016                698.0      4390.0

### Step 3: Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

> **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`

### Step 4: Style and format the line plot to ensure a professionally styled visualization.


In [38]:
# Plot prices_square_foot_by_year. 
# Inclued labels for the x- and y-axes, and a title
prices_square_foot_by_year.hvplot.line(
    title="Sales Price Per Square Foot And Average Gross Rent",
    xlabel="Year", 
    ylabel="Gross Rent", 
    figsize= (20,10)
)

:NdOverlay   [Variable]
   :Curve   [year]   (value)

### Step 6: Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

**Question:** Did any year experience a drop in the average sale price per square foot compared to the previous year?

**Answer:** # Yes. In year 2011, there was price per square footdrop, copared to 2010. 

**Question:** If so, did the gross rent increase or decrease during that year?

**Answer:** # The gross rent increased in 2011 vs 2010.


---

### Step 1: Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

In [39]:
# Group by year and neighborhood and then create a new dataframe of the mean values
prices_by_year_by_neighborhood = sfo_data_df.groupby(['year','neighborhood']).mean()

# Review the DataFrame
prices_by_year_by_neighborhood


sale_price_sqr_foot  housing_units  gross_rent
year neighborhood                                                          
2010 Alamo Square                     291.182945       372560.0      1239.0
     Anza Vista                       267.932583       372560.0      1239.0
     Bayview                          170.098665       372560.0      1239.0
     Buena Vista Park                 347.394919       372560.0      1239.0
     Central Richmond                 319.027623       372560.0      1239.0
...                                          ...            ...         ...
2016 Telegraph Hill                   903.049771       384242.0      4390.0
     Twin Peaks                       970.085470       384242.0      4390.0
     Van Ness/ Civic Center           552.602567       384242.0      4390.0
     Visitacion Valley                328.319007       384242.0      4390.0
     Westwood Park                    631.195426       384242.0      4390.0

[392 rows x 3 columns]

### Step 2: Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

In [40]:
# Filter out the housing_units
prices_by_year_by_neighborhood = prices_by_year_by_neighborhood.drop(columns=['housing_units'])

# Review the first and last five rows of the DataFrame
display(prices_by_year_by_neighborhood.head())
display(prices_by_year_by_neighborhood.tail())

sale_price_sqr_foot  gross_rent
year neighborhood                                     
2010 Alamo Square               291.182945      1239.0
     Anza Vista                 267.932583      1239.0
     Bayview                    170.098665      1239.0
     Buena Vista Park           347.394919      1239.0
     Central Richmond           319.027623      1239.0

sale_price_sqr_foot  gross_rent
year neighborhood                                           
2016 Telegraph Hill                   903.049771      4390.0
     Twin Peaks                       970.085470      4390.0
     Van Ness/ Civic Center           552.602567      4390.0
     Visitacion Valley                328.319007      4390.0
     Westwood Park                    631.195426      4390.0

### Step 3: Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

### Step 4: Style and format the line plot to ensure a professionally styled visualization.

In [41]:
# Use hvplot to create an interactive line plot of the average price per square foot
# The plot should have a dropdown selector for the neighborhood
prices_by_year_by_neighborhood.hvplot.line(groupby='neighborhood')

:DynamicMap   [neighborhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

### Step 6: Use the interactive visualization to answer the following question:

**Question:** For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

**Answer:** # The average Price per square foot for 2016 is 88.40 which is lower than 2012 price per square for at 344.49.

---

### Step 1: Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

In [42]:
# Load neighborhoods coordinates data
neighborhood_locations_csv= Path('Resources/neighborhoods_coordinates.csv')
neighborhood_locations_df = pd.read_csv(neighborhood_locations_csv)
#clean dataframes of any nulls 
neighborhood_locations_df.isnull()
neighborhood_locations_df.dropna()

neighborhood_locations_df.isnull().sum()


Neighborhood    0
Lat             0
Lon             0
dtype: int64

In [43]:
#Review the DataFrame
display(neighborhood_locations_df)

Neighborhood        Lat         Lon
0         Alamo Square  37.791012 -122.402100
1           Anza Vista  37.779598 -122.443451
2              Bayview  37.734670 -122.401060
3      Bayview Heights  37.728740 -122.410980
4       Bernal Heights  37.728630 -122.443050
..                 ...        ...         ...
68         West Portal  37.740260 -122.463880
69    Western Addition  37.792980 -122.435790
70  Westwood Highlands  37.734700 -122.456854
71       Westwood Park  37.734150 -122.457000
72         Yerba Buena  37.792980 -122.396360

[73 rows x 3 columns]

In [44]:
neighborhood_locations_df.duplicated()

0     False
1     False
2     False
3     False
4     False
      ...  
68    False
69    False
70    False
71    False
72    False
Length: 73, dtype: bool

In [45]:
neighborhood_locations_df.duplicated().sum()


0

In [46]:
#clean Data by dropping all duplicates 
#neighborhood_locations_df=neighborhood_locations_df.drop_duplicates()

#neighborhood_locations_df.set_index('Neighborhood', inplace=True)
#neighborhood_locations_df.set_index('Neighborhood')
#clean Data by dropping all duplicates
neighborhood_locations_df=neighborhood_locations_df.drop_duplicates()
#neighborhood_locations_df.set_index('Neighborhood', inplace=True)
neighborhood_locations_df=neighborhood_locations_df.set_index('Neighborhood')

In [47]:
display(neighborhood_locations_df)

Lat         Lon
Neighborhood                             
Alamo Square        37.791012 -122.402100
Anza Vista          37.779598 -122.443451
Bayview             37.734670 -122.401060
Bayview Heights     37.728740 -122.410980
Bernal Heights      37.728630 -122.443050
...                       ...         ...
West Portal         37.740260 -122.463880
Western Addition    37.792980 -122.435790
Westwood Highlands  37.734700 -122.456854
Westwood Park       37.734150 -122.457000
Yerba Buena         37.792980 -122.396360

[73 rows x 2 columns]

### Step 2: Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

In [48]:
# Calculate the mean values for each neighborhood
all_neighborhood_info_df = round(sfo_data_df.groupby('neighborhood').mean())


# Review the resulting DataFrame

all_neighborhood_info_df.isnull()
#all_neighborhood_info_df.dropna(inplace=True)
#all_neighborhood_info_df.isnull()


display(all_neighborhood_info_df)

year  sale_price_sqr_foot  housing_units  gross_rent
neighborhood                                                              
Alamo Square        2013.0                366.0       378401.0      2817.0
Anza Vista          2013.0                373.0       379050.0      3032.0
Bayview             2012.0                205.0       376454.0      2318.0
Bayview Heights     2015.0                591.0       382295.0      3739.0
Bernal Heights      2014.0                577.0       379374.0      3080.0
...                    ...                  ...            ...         ...
West Portal         2012.0                498.0       376941.0      2516.0
Western Addition    2012.0                308.0       377428.0      2555.0
Westwood Highlands  2012.0                534.0       376454.0      2250.0
Westwood Park       2015.0                687.0       382295.0      3959.0
Yerba Buena         2012.0                577.0       377428.0      2555.0

[73 rows x 4 columns]

### Step 3: Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. 

Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. 

The second cell cleans the data and sets the “Neighborhood” column. 

Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

In [49]:
# Using the Pandas `concat` function, join the 
# neighborhood_locations_df and the all_neighborhood_info_df DataFrame
# The axis of the concatenation is "columns".
# The concat function will automatially combine columns with
# identical information, while keeping the additional columns.
all_neighborhoods_df = pd.concat(
    [neighborhood_locations_df, all_neighborhood_info_df], 
    axis="columns",join='inner',
    sort=False
)

# Review the resulting DataFrame
#display(all_neighborhoods_df.head())
#display(all_neighborhoods_df.tail())
display(all_neighborhoods_df)
#display(all_neighborhood_info_df)

Lat         Lon    year  sale_price_sqr_foot  \
Alamo Square        37.791012 -122.402100  2013.0                366.0   
Anza Vista          37.779598 -122.443451  2013.0                373.0   
Bayview             37.734670 -122.401060  2012.0                205.0   
Bayview Heights     37.728740 -122.410980  2015.0                591.0   
Buena Vista Park    37.768160 -122.439330  2013.0                453.0   
...                       ...         ...     ...                  ...   
West Portal         37.740260 -122.463880  2012.0                498.0   
Western Addition    37.792980 -122.435790  2012.0                308.0   
Westwood Highlands  37.734700 -122.456854  2012.0                534.0   
Westwood Park       37.734150 -122.457000  2015.0                687.0   
Yerba Buena         37.792980 -122.396360  2012.0                577.0   

                    housing_units  gross_rent  
Alamo Square             378401.0      2817.0  
Anza Vista               379050.0      3032.0  
Bayview                  376454.0      2318.0  
Bayview Heights          382295.0      3739.0  
Buena Vista Park         378076.0      2699.0  
...                           ...         ...  
West Portal              376941.0      2516.0  
Western Addition         377428.0      2555.0  
Westwood Highlands       376454.0      2250.0  
Westwood Park            382295.0      3959.0  
Yerba Buena              377428.0      2555.0  

[69 rows x 6 columns]

In [50]:
# Call the dropna function to remove any neighborhoods that do not have data
all_neighborhoods_df = all_neighborhoods_df.reset_index().dropna()

# Rename the "index" column as "Neighborhood" for use in the Visualization
all_neighborhoods_df = all_neighborhoods_df.rename(columns={"index": "Neighborhood"})


# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())

Neighborhood        Lat         Lon    year  sale_price_sqr_foot  \
0      Alamo Square  37.791012 -122.402100  2013.0                366.0   
1        Anza Vista  37.779598 -122.443451  2013.0                373.0   
2           Bayview  37.734670 -122.401060  2012.0                205.0   
3   Bayview Heights  37.728740 -122.410980  2015.0                591.0   
4  Buena Vista Park  37.768160 -122.439330  2013.0                453.0   

   housing_units  gross_rent  
0       378401.0      2817.0  
1       379050.0      3032.0  
2       376454.0      2318.0  
3       382295.0      3739.0  
4       378076.0      2699.0

Neighborhood       Lat         Lon    year  sale_price_sqr_foot  \
64         West Portal  37.74026 -122.463880  2012.0                498.0   
65    Western Addition  37.79298 -122.435790  2012.0                308.0   
66  Westwood Highlands  37.73470 -122.456854  2012.0                534.0   
67       Westwood Park  37.73415 -122.457000  2015.0                687.0   
68         Yerba Buena  37.79298 -122.396360  2012.0                577.0   

    housing_units  gross_rent  
64       376941.0      2516.0  
65       377428.0      2555.0  
66       376454.0      2250.0  
67       382295.0      3959.0  
68       377428.0      2555.0

### Step 4: Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

* Set the `geo` parameter to True.
* Set the `size` parameter to “sale_price_sqr_foot”.
* Set the `color` parameter to “gross_rent”.
* Set the `frame_width` parameter to 700.
* Set the `frame_height` parameter to 500.
* Include a descriptive title.

In [52]:
# Create a plot to analyze neighborhood info
# Create a plot to analyze neighborhood info
map_plot = all_neighborhoods_df.hvplot.points(
    'Lon',
    'Lat',
    geo=True,
    tiles='OSM',
    color='gross_rent',
    size = 'sale_price_sqr_foot',
    #scale=.04,
    frame_width = 700,
    frame_height = 500
    )
map_plot

ImportError: Geographic projection support requires GeoViews and cartopy.

### Step 5: Use the interactive map to answer the following question:

**Question:** Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

**Answer:** # Based on the interactive map, Westwood Neighbourhood has the highest gross rent and as well as the highest sale price 
per square foot.

## Compose Your Data Story

Based on the visualizations that you have created, compose a data story that synthesizes your analysis by answering the following questions:

**Question:**  How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

**Answer:** # The trend in rental income across the city of San Fransisco is very much consistent and on a rise. where as price per square foor does fluctuate through out the city with some on up trend and some on the down trend as of 2016. 

**Question:** What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

**Answer:** # In terms of real estate investment for an income property I would recommend Siver Terrace, as it is priced at 170 dollar per square foot with a gross monthly rent of 3528 dollars. 